tests/client.spec.js

```diff
@@ -91,6 +89,120 @@ describe(`client`, () => {
+       * will never run.
+       */
+      describe('unsubscribe from responseStream', () => {
+        it('via take operator', done => {
+          const { impl, server } = initServerPayload;
+          const callObs = makeCall(false, 10);
+
+          let nextCalls = 0;
+          const expectedCalls = 2;
+
+          callObs.take(expectedCalls).subscribe({
+            next: () => {
+              expect(impl.sayMultiHello.holdingObservers.size).to.be.eql(1);
+              nextCalls++;
+            },
+            error: maybeError => {
+              done(maybeError);
+            },
+            complete: () => {
+              setTimeout(() => {
+                expect(nextCalls).to.be.eql(expectedCalls);
+                expect(impl.sayMultiHello.holdingObservers.size).to.be.eql(0);
+                // This next line is key, if stream is still open cancelCache will have length.
+                // Checking its 0 ensures we've closed stream for good.
+                expect(grpcAPI.cancelCache.size).to.be.eql(0);
+                done();
+              }, 50);
+              conn.close();
+              server.forceShutdown();
+            }
+          });
+        });
+        it('manually', done => {
+          const { impl, server } = initServerPayload;
+          const callObs = makeCall(false, 10);
+
+          let nextCalls = 0;
+          const expectedCalls = 4;
+
+          const ret = callObs.subscribe({
+            next: () => {
+              expect(impl.sayMultiHello.holdingObservers.size).to.be.eql(1);
+              nextCalls++;
+              if (nextCalls === expectedCalls) {
+                ret.unsubscribe();
+              }
+              if (nextCalls > expectedCalls) {
+                throw new Error('Too many next calls');
+              }
+            },
+            error: maybeError => {
+              done(maybeError);
+            },
+            complete: () => {
+              done(new Error('Unsubscribe should not run complete'));
+            }
+          });
+          // add RXJS teardown logic called on unsubscribe
+          ret.add(() => {
+            conn.close();
+            server.forceShutdown();
+            setTimeout(() => {
+              expect(nextCalls).to.be.eql(expectedCalls, 'nextCalls eql expectedCalls');
+              expect(impl.sayMultiHello.holdingObservers.size).to.be.eql(0, 'holdingObservers eql 0');
+              expect(grpcAPI.cancelCache.size).to.be.eql(0, 'cancelCache eql 0');
+              done();
+            }, 50); // give test connection time to shut down
+          });
+        });
+        it('manually after delay so all responseStreams have come in', done => {
+          const { impl, server } = initServerPayload;
+          const callObs = makeCall(false, 10);
+
+          let nextCalls = 0;
+          const delay = 100;
+          const expectedCalls = 4;
+
+          const ret = callObs.delay(delay).subscribe({
```

Comment:
Weird I tried this and never got this working. So I gave up and passed the delay to the server directly.

Reply:
I think it's testing a different case... what is tested here is

```
observer A subscribes with 100 ms delay and expected messages 2
client/ server makes request for 10 messages
responseStream fires 10 times, adds to observer.queue
meanwhile, after 100ms
observer A processes queue, takes 2 messages
messages 3-10 result in error which I was only able to get around with call.on('error', () => {})
```

I think the test you have which moves delay to the server does something like this: 

```
observer A subscribes
client/ server makes request for 10 messages with 100ms delay
after 100ms
responseStream fires 1 time
after 100ms
responseStream fires 1 time
...
```

Seems like both are valid cases to test.